### PR TITLE
feat: add deposits for DIP provider pallet

### PR DESCRIPTION
**Pull request by ntn-x2**
Fixes https://github.com/KILTprotocol/ticket/issues/2983.

Add a generic pallet to store deposits, as well as a hooking mechanism to both the existing provider pallet (to allow runtimes that take deposits to take deposits, or to do any other operation), and to the new deposit pallet, which allows other pallets to clean up storage if a storage is reclaimed back by its owner (e.g., an identity commitment is removed).

Tests will follow in a separate PR, when time allows for them 😁 The whole DIP branch won't be merged on `develop` before tests are written, so it's not a risk. The logic has been tested by spinning up a local network and trying to commit an identity (which takes a deposit), remove the commitment (which frees up the deposit), re-commit the identity, reclaim the deposit (which removes the commitment).

- [x] Implement logic for deposit removal hooks
- [x] Test
- [x] Review

### CODE REVIEW

File: dip-template/runtimes/dip-provider/src/dip.rs [hunk 1]
```diff
@@ -27,22 +27,59 @@ use runtime_common::dip::{
...
-#[derive(Encode, Decode, TypeInfo)]
-pub struct RuntimeApiDipProofRequest {
-	pub(crate) identifier: DidIdentifier,
-	pub(crate) version: IdentityCommitmentVersion,
-	pub(crate) keys: Vec<KeyIdOf<Runtime>>,
-	pub(crate) accounts: Vec<LinkableAccountId>,
-	pub(crate) should_include_web3_name: bool,
+use crate::{AccountId, Balances, DidIdentifier, Hash, Runtime, RuntimeEvent};
+
+pub mod runtime_api {
```
Comment (ntn-x2) [inline dip-template/runtimes/dip-provider/src/dip.rs:32]: Please ignore this, not relevant for what I'd like to get reviewed.

File: dip-template/runtimes/dip-provider/src/dip.rs [hunk 2]
```diff
@@ -27,22 +27,59 @@ use runtime_common::dip::{
...
+
+	pub struct Namespace;
+
+	impl Get<BoundedVec<u8, ConstU32<MAX_NAMESPACE_LENGTH>>> for Namespace {
+		fn get() -> BoundedVec<u8, ConstU32<MAX_NAMESPACE_LENGTH>> {
+			(*b"DipProvider").to_vec().try_into().expect("Should never fail.")
+		}
+	}
+
+	pub const DEPOSIT_AMOUNT: Balance = 100 * UNIT;
```
Comment (ntn-x2) [inline dip-template/runtimes/dip-provider/src/dip.rs:70]: This will be a constant specified somewhere in the runtime config.

Comment (trusch) [reply dip-template/runtimes/dip-provider/src/dip.rs:70]: If we want to reuse this pallet, we should put some more thought in how to configure it for different actions that require deposits.

File: dip-template/runtimes/dip-provider/Cargo.toml
Comment (Ad96el) [inline dip-template/runtimes/dip-provider/Cargo.toml:1]: Maybe not part of this PR, but does it not make more sense to have the dip-template in his own repo?

File: dip-template/runtimes/dip-provider/src/lib.rs [hunk 3]
```diff
@@ -568,14 +569,14 @@ impl_runtime_apis! {
...
+		fn generate_proof(request: runtime_api::DipProofRequest) -> Result<CompleteMerkleProof<Hash, DidMerkleProofOf<Runtime>>, runtime_api::DipProofError> {
 			let linked_did_info = match <Runtime as pallet_dip_provider::Config>::IdentityProvider::retrieve(&request.identifier) {
 				Ok(Some(linked_did_info)) => Ok(linked_did_info),
-				Ok(None) => Err(RuntimeApiDipProofError::IdentityNotFound),
-				Err(e) => Err(RuntimeApiDipProofError::IdentityProviderError(e))
+				Ok(None) => Err(runtime_api::DipProofError::IdentityNotFound),
+				Err(e) => Err(runtime_api::DipProofError::IdentityProviderError(e))
 			}?;
-			DidMerkleRootGenerator::<Runtime>::generate_proof(&linked_did_info, request.version, request.keys.iter(), request.should_include_web3_name, request.accounts.iter()).map_err(RuntimeApiDipProofError::MerkleProofError)
+			DidMerkleRootGenerator::<Runtime>::generate_proof(&linked_did_info, request.version, request.keys.iter(), request.should_include_web3_name, request.accounts.iter()).map_err(runtime_api::DipProofError::MerkleProofError)
```
Comment (Ad96el) [inline dip-template/runtimes/dip-provider/src/lib.rs:579]: These lines look very complicated. Can we maybe split it? Especially if it is used as template

**Review by Ad96el — COMMENTED** [review]
Very nice! I like the architectural design and the default implementation of the traits. Some comments are a bit nitpicking - sorry for that. I assume the pallet-deposit-storage will also be deployed on Spiritnet, right? If so, the storage version has to be initialized in the runtimes. Additionally, I would appreciate more documentation. Perhaps this can be addressed in another PR?

File: pallets/pallet-deposit-storage/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		},
+	};
+	use frame_system::pallet_prelude::*;
+	use parity_scale_codec::FullCodec;
+	use scale_info::TypeInfo;
+	use sp_runtime::DispatchError;
+	use sp_std::fmt::Debug;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-deposit-storage/src/lib.rs:51]: It's storage version 0, so migrations would not fail if they check that, right? The default would still be 0. I would migrate it once it is bumped to 1, no?

File: pallets/pallet-deposit-storage/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	pub trait Config: frame_system::Config {
+		#[pallet::constant]
+		type MaxKeyLength: Get<u32>;
+
+		type CheckOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+		type Currency: Mutate<Self::AccountId, Reason = Self::RuntimeHoldReason>;
+		type DepositHooks: DepositStorageHooks<Self>;
+		type Namespace: Parameter;
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type RuntimeHoldReason: From<HoldReason> + Clone + PartialEq + Debug + FullCodec + MaxEncodedLen + TypeInfo;
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-deposit-storage/src/lib.rs:68]: Yeah I haven't worked on the docs yet. Sorry if this is hard to read, I will work on a docs PR later on.

File: pallets/pallet-dip-provider/src/lib.rs [hunk 6]
```diff
@@ -31,23 +31,17 @@ pub mod pallet {
 	use frame_support::{pallet_prelude::*, traits::EnsureOrigin};
 	use frame_system::pallet_prelude::*;
 	use parity_scale_codec::FullCodec;
-	use sp_io::MultiRemovalResults;
 	use sp_std::fmt::Debug;
 
-	use crate::traits::{IdentityCommitmentGenerator, IdentityProvider, SubmitterInfo};
+	use crate::traits::{IdentityCommitmentGenerator, IdentityProvider, ProviderHooks, SubmitterInfo};
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/lib.rs:36]: You mean that there is a `Config` type which is also called `ProviderHooks`?

File: dip-template/runtimes/dip-provider/Cargo.toml
Comment (ntn-x2) [reply dip-template/runtimes/dip-provider/Cargo.toml:1]: Yes, absolutely. Not part of this PR, but it will be its own repo, eventually.

File: dip-template/runtimes/dip-provider/src/lib.rs (hunk 3 above)
Comment (ntn-x2) [reply dip-template/runtimes/dip-provider/src/lib.rs]: Like this? https://github.com/KILTprotocol/kilt-node/pull/574/commits/24cb5ac5c19b80461a0d86ba6515b8c521aa255a

File: pallets/pallet-deposit-storage/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-deposit-storage/src/lib.rs:51]: Are you sure that the on-chain storage version is set to 0 by default for a new pallet? As far as I know, the on-chain version needs to be initialized. This is what we did in this [PR](https://github.com/KILTprotocol/kilt-node/commit/8e36fb26011887f726eaef21036cf56e02450c4a#diff-0fb85acd71505115281e23edf6e01a99b84b52eb8ba07629ba7518534a71d3d7R54) for the multisig pallet.

[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-deposit-storage/src/lib.rs:51]: No I'm not saying it is set to 0, but when being read, if not present, the default value of `0` is returned, which matches the pallet storage version, hence the check `stored_version == pallet_version` will not fail. Of course, once the pallet version is updated to `1`, then it will have to be written on storage. The issue with Substrate pallet, is that it was not written on storage when the pallet version was greater than `0`, and THAT is a problem.

File: pallets/pallet-dip-provider/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-dip-provider/src/lib.rs:36]: No, sorry - I'm not referring to the Config. I shouldn't have marked it here.

In this [line](https://github.com/KILTprotocol/kilt-node/pull/574/files#diff-6dc5259ea5e53a98cd0b1ecd6ce9819c2efbb44cb8957070cb0e69b75fbdfa65R65), you're implementing the trait `ProviderHooks` for `FixedDepositCollectorViaDepositsPallet`. When I read the line, my initial assumption was that `ProviderHooks` are from the pallet-deposit-storage. It would be more clear if the trait somehow indicated that the hooks are from the DIP pallet.

File: pallets/pallet-deposit-storage/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-deposit-storage/src/lib.rs:51]: Ah, I see. There is a default implementation. Thanks for your explanation :racehorse:

File: dip-template/runtimes/dip-provider/src/lib.rs (hunk 3 above)
Comment (Ad96el) [reply dip-template/runtimes/dip-provider/src/lib.rs]: Yes, I think it is already better. Maybe we can have a type `IdentityProviderOf<Runtime>` instead of `<Runtime as pallet_dip_provider::Config>::IdentityProvider`. But this is not really urgent or important.

File: pallets/pallet-dip-provider/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/lib.rs:36]: Mhh what makes you think that `ProviderHooks` is part of the deposit pallet? Because the type implementing it is declared here? Also, the trait is imported from `pallet_dip_provider` with `use pallet_dip_provider::traits::ProviderHooks`. But I anyway renamed the import to avoid confusion, what do you think?  https://github.com/KILTprotocol/kilt-node/pull/574/commits/f73778daaf1dbf38db5d0a6a948698c124001146

File: dip-template/runtimes/dip-provider/src/lib.rs (hunk 3 above)
Comment (ntn-x2) [reply dip-template/runtimes/dip-provider/src/lib.rs]: Yup: https://github.com/KILTprotocol/kilt-node/pull/574/commits/f73778daaf1dbf38db5d0a6a948698c124001146.

File: pallets/pallet-dip-provider/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-dip-provider/src/lib.rs:36]: I believe that 'ProviderHooks' is a broad term that could be associated with many pallets. I think that with the renaming, it becomes more declarative. I needed to scroll up to see where the trait comes from - will be not necessary anymore.

**Review by Ad96el — APPROVED** [review]
:rocket:

File: dip-template/runtimes/dip-provider/src/dip.rs (hunk 2 above)
Comment (ntn-x2) [reply dip-template/runtimes/dip-provider/src/dip.rs]: As discussed on Slack, the initial scope of this pallet is to support storing/removing deposits. More complex logic COULD be added in the future, if we see fit. For now I'm marking this conversation as resolved.

File: pallets/pallet-dip-provider/src/lib.rs [hunk 7]
```diff
@@ -31,23 +31,18 @@ pub mod pallet {
...
 
+	pub type IdentityProviderOf<T> = <T as Config>::IdentityProvider;
 	pub type IdentityOf<T> = <<T as Config>::IdentityProvider as IdentityProvider<<T as Config>::Identifier>>::Success;
 	pub type IdentityCommitmentVersion = u16;
 
 	pub const LATEST_COMMITMENT_VERSION: IdentityCommitmentVersion = 0;
+	pub const MAX_COMMITMENTS_PER_IDENTITY: u16 = LATEST_COMMITMENT_VERSION + 1;
```
Comment (Ad96el) [inline pallets/pallet-dip-provider/src/lib.rs:43]: What is the purpose of this constant? It appears that it may not be used.

Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/lib.rs]: Ops, good catch! I removed it. It was meant to put an upper bound on the number of commitments per user, but since it's a double map with (DID + commitment version), there is already an upper bound given by the number of versions supported.

Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/lib.rs]: https://github.com/KILTprotocol/kilt-node/pull/574/commits/1841f0fa64e49381d4e18d3124950624a8343d65

File: pallets/pallet-dip-provider/src/lib.rs [hunk 8]
```diff
@@ -31,23 +31,17 @@ pub mod pallet {
...
 	use parity_scale_codec::FullCodec;
-	use sp_io::MultiRemovalResults;
 	use sp_std::fmt::Debug;
 
-	use crate::traits::{IdentityCommitmentGenerator, IdentityProvider, SubmitterInfo};
+	use crate::traits::{IdentityCommitmentGenerator, IdentityProvider, ProviderHooks, SubmitterInfo};
 
+	pub type IdentityProviderOf<T> = <T as Config>::IdentityProvider;
 	pub type IdentityOf<T> = <<T as Config>::IdentityProvider as IdentityProvider<<T as Config>::Identifier>>::Success;
 	pub type IdentityCommitmentVersion = u16;
```
Comment (Ad96el) [inline pallets/pallet-dip-provider/src/lib.rs:40]: For my learning: Why not create a type in the config that implements the default trait? Wouldn't it be cleaner? I often struggle to determine what should be included in the configuration and what should not. :confused:

Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/lib.rs:40]: You mean using the latest Substrate features that allow to specify defaults for a pallet's `Config` trait? Or what do you mean?

File: pallets/pallet-deposit-storage/src/deposit.rs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+			FixedDepositCollectorViaDepositsPalletError::DepositAlreadyTaken => 0,
+			FixedDepositCollectorViaDepositsPalletError::DepositNotFound => 1,
+			FixedDepositCollectorViaDepositsPalletError::FailedToHold => 2,
+			FixedDepositCollectorViaDepositsPalletError::FailedToRelease => 3,
+			FixedDepositCollectorViaDepositsPalletError::Internal => u16::MAX,
+		}
+	}
+}
+
+impl<Runtime, DepositsNamespace, FixedDepositAmount> DipProviderHooks<Runtime>
```
Comment (trusch) [inline pallets/pallet-deposit-storage/src/deposit.rs:65]: Somehow it feels weird to see this tight coupling of the deposits and the dip pallet here. Shouldn't this logic also be passed in while configuring the pallet for a specific runtime like with the removal?

**Review by trusch — APPROVED** [review]
lgtm in general, just one question/remark

File: pallets/pallet-deposit-storage/src/deposit.rs (hunk 9 above)
Comment (ntn-x2) [reply pallets/pallet-deposit-storage/src/deposit.rs:65]: Do you think the name of the type is misleading? This is already a concrete implementation of a trait of the DIP provider pallet, and it is supposed to work only when you are storing identity commitments and want to take deposits for them. I don't understand what you refer with the logic being passed? Can you elaborate on that?

Comment (trusch) [reply pallets/pallet-deposit-storage/src/deposit.rs:65]: I think the question is more from an architectural point of view: Why does something that is obviously specific to the DIP pallet (like this implementation of the DipProviderHooks) living within the deposit-storage pallet? Consequently this would mean that the deposit pallet code base would need to grow with every new "consumer" pallet that uses the deposit pallet. I think no action item should result from this comment, I'm only interested in your reasoning for structuring it this way.

Comment (ntn-x2) [reply pallets/pallet-deposit-storage/src/deposit.rs:65]: The problem is the Rust orphan rule. You can only implement a trait for a type if either of them is native to the crate. In this case, it would mean either implementing the trait in the pallet storage or in the pallet provider crate, as I see no point in having it in any other place. Would you rather have the type in the provider crate?

File: pallets/pallet-dip-provider/src/lib.rs (hunk 8 above)
Comment (ntn-x2) [reply pallets/pallet-dip-provider/src/lib.rs:40]: I took a shot in https://github.com/KILTprotocol/kilt-node/pull/577. Resolving this, let me know there what you think!
